disk/src/com/treode/disk/edit/DriveGroup.scala

```diff
@@ -221,15 +225,25 @@ private class DriveGroup (
       // - make the drive, and
       // - add it to the list of new attaches.
       for (a <- change.attaches) {
-        if (attached contains a.path)
-          throw new IllegalArgumentException (s"Already attached ${quote (a.path)}")
-        if (attaching contains a.path)
-          throw new IllegalArgumentException (s"Already attaching ${quote (a.path)}")
-        attaching += a.path
-        val file = files.open (a.path, READ, WRITE)
-        val drive = new Drive (file, a.geometry, false, dno, a.path)
-        dno += 1
-        newAttaches ::= drive
+        if (attached contains a.path) {
+          errors.add (AlreadyAttached(a.path))
+        } else if (attaching contains a.path) {
+          errors.add (AlreadyAttaching(a.path))
+        } else {
+          attaching += a.path
+          val file = files.open (a.path, READ, WRITE)
+          val drive = new Drive (file, a.geometry, false, dno, a.path)
+          dno += 1
+          newAttaches ::= drive
+        }
+      }
+
+      // If there are attachment errors, close all opened files and pass errors.
+      if (errors.hasErrors) {
```

Comment:
This invokes the callback, and it also falls through and enqueues the callback in `changers` to be invoked later. Also, it comes before the check for drain errors. Be sure to add drain errors to the Notification. After than, check for errors and either invoke the callback now, or put it into the `changers` list, but not both.